crates/uv-publish/src/lib.rs

```diff
@@ -0,0 +1,789 @@
+
+    add_vec("classifiers", metadata.classifiers);
+    add_vec("platform", metadata.platforms);
+    add_vec("requires_dist", metadata.requires_dist);
+    add_vec("provides_dist", metadata.provides_dist);
+    add_vec("obsoletes_dist", metadata.obsoletes_dist);
+    add_vec("requires_external", metadata.requires_external);
+    add_vec("project_urls", metadata.project_urls);
+
+    Ok(form_metadata)
+}
+
+async fn build_request(
+    file: &Path,
+    filename: &DistFilename,
+    registry: &Url,
+    client: &BaseClient,
+    username: Option<&str>,
+    password: Option<&str>,
+    form_metadata: Vec<(&'static str, String)>,
+) -> Result<RequestBuilder, PublishPrepareError> {
+    let mut form = reqwest::multipart::Form::new();
+    for (key, value) in form_metadata {
+        form = form.text(key, value);
+    }
+
+    let file: tokio::fs::File = fs_err::tokio::File::open(file).await?.into();
+    let file_reader = Body::from(file);
+    form = form.part(
+        "content",
+        Part::stream(file_reader).file_name(filename.to_string()),
+    );
+
+    let url = if let Some(username) = username {
+        if password.is_none() {
+            // Attach the username to the URL so the authentication middleware can find the matching
+            // password.
+            let mut url = registry.clone();
+            let _ = url.set_username(username);
+            url
+        } else {
+            // We set the authorization header below.
+            registry.clone()
+        }
+    } else {
+        registry.clone()
+    };
+
+    let mut request = client
+        .client()
+        .post(url)
+        .multipart(form)
+        // Ask PyPI for a structured error messages instead of HTML-markup error messages.
+        // For other registries, we ask them to return plain text over HTML. See
+        // [`PublishSendError::extract_remote_error`].
+        .header(
+            reqwest::header::ACCEPT,
+            "application/json;q=0.9, text/plain;q=0.8, text/html;q=0.7",
+        );
+    if let (Some(username), Some(password)) = (username, password) {
+        debug!("Using username/password basic auth");
+        let credentials = BASE64_STANDARD.encode(format!("{username}:{password}"));
+        request = request.header(AUTHORIZATION, format!("Basic {credentials}"));
+    }
+    Ok(request)
+}
+
+/// Returns `true` if the file was newly uploaded and `false` if it already existed.
+async fn handle_response(registry: &Url, response: Response) -> Result<bool, PublishSendError> {
+    let status_code = response.status();
+    debug!("Response code for {registry}: {status_code}");
+    trace!("Response headers for {registry}: {response:?}");
+
+    // TODO(konsti): There's some strange behavior here we're not handling yet: When I POST to
+    // https://test.pypi.org/simple/, I get a method not allowed error (as it should be), but when
+    // I post to https://test.pypi.org/simple (no slash) it returns a 200 with the content of the
```

Comment:
This is likely reqwest doing a POST-redirect-to-GET after seeing a 301, which I think is the proper behavior.
It smells like a backend bug to redirect an unexpected POST in this way though.